flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ResourceEventHandler.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.resourcemanager.active;
+
+import org.apache.flink.runtime.clusterframework.types.ResourceID;
+import org.apache.flink.runtime.clusterframework.types.ResourceIDRetrievable;
+
+import java.util.Collection;
+
+/**
+ * Callback interfaces for handling resource events from external resource managers.
+ */
+public interface ResourceEventHandler<WorkerType extends ResourceIDRetrievable> {
+
+	/**
+	 * Notifies that workers of previous attempt have been recovered from the external resource manager.
+	 *
+	 * @param recoveredWorkers Collection of worker nodes, in the deployment specific type.
+	 */
+	void onPreviousAttemptWorkersRecovered(Collection<WorkerType> recoveredWorkers);
+
+	/**
+	 * Notifies that the worker has been terminated.
+	 *
+	 * <p>See also {@link ResourceManagerDriver#requestResource}.
+	 *
+	 * @param resourceId Identifier of the terminated worker.
+	 */
+	void onWorkerTerminated(ResourceID resourceId);
+
+	/**
+	 * Notifies that an error has occurred that the process cannot proceed.
+	 *
+	 * @param exception Exception that describes the error.
+	 */
+	void onError(Throwable exception);
+
+	/**
+	 * Execute given runnable in the rpc main thread.
+	 *
+	 * @param runnable Runnable to be executed.
+	 */
+	void handleInMainThread(Runnable runnable);
```

Comment:
Would it make sense to pass into the `ResourceManagerDriver.initialize()` a `MainThreadExecutor` instead of offering this method? The advantage would be that once could directly run future call backs in the main thread and does not have to call this method from the callback. Moreover, it would separate the concerns a bit better because the event handler does not need to know about the main thread if I'm not mistaken.

Reply:
I was also hesitating between the current approach and the one you described. The problem is that, `MainThreadExecutor` is a protected inner class of `RpcEndpoint`. I'm not entirely sure whether we want to break that visibility limit for this purpose. WDYT?

Reply:
I think we could pass in the `ComponentMainThreadExecutor` which is a public interface.

Reply:
We might even be able to pass in an `Executor` if we don't need any other methods than `execute`.

Reply:
I recalled the other reason I did not pass in the main thread executor. It has been two weeks.

I did tried passing in `ComponentMainThreadExecutor` but it doesn't work. Because `FencedRpcEndpoint#fencedMainThreadExecutor` could be changed via `setFencingToken`. So the main thread executor at the time initializing will no longer be the main thread executor after RM accepting the leadership.

Reply:
In this case, one could update the `ResourceManagerDriver` whenever the `ResourceManager` obtains the leadership.

On a side note, I believe that we should actually not reuse the `ResourceManager` across leader sessions and instead create a new one if the RM gets the leadership (similar to what we do with the `Dispatcher` nowadays).

Reply:
If we update the `ResourceManagerDriver` whenever the `ResourceManager` obtains the leadership, we are hurting another separation of concerns, because `ResourceManager` will need to understand when the main thread executor will be changed, which is currently hidden inside `FencedRpcEndpoint`.

I had a feeling that it could be fragile to maintain the consistency between the two main thread executors.

Reply:
I think this is already the case: See `ResourceManager.grantLeadership`, `ResourceManager.tryAcceptLeadership` and `ResourceManager.startServicesOnLeadership`.

Reply:
There's probably an easier way to keep the main thread executor on driver side up to date.

We can introduce the following implementation of `Executor` as an inner class of `ActiveResourceManager`, and pass it into `ResourceManagerDriver.initialize()`.

```
/**
 * Always execute on the current main thread executor.
 */
private class GatewayExecutor implements Executor {
	@Override
	public void execute(Runnable command) {
		getMainThreadExecutor().execute(command);
	}
}
```

This should be equivalent to currently calling `ResourceEventHandler. handleInMainThread()`, while provide better separation of concerns. WDYT?

Reply:
And this could be simplified with lambda.
```
resourceManagerDriver.initialize(this, command -> getMainThreadExecutor().execute(command));
```